src/core/services/shift_service.py

```diff
@@ -205,3 +210,28 @@ async def list_all_shifts(
 
     async def get_open_for_registration_shift_id(self) -> UUID:
         return await self.__shift_repository.get_open_for_registration_shift_id()
+
+    async def cancel_shift(
+        self, bot: Application, id: UUID, cancel_shift_data: Optional[ShiftCancelRequest] = None
+    ) -> Shift:
+        shift = await self.__shift_repository.get_shift_with_request(id)
+        final_message = "Смена отменена"
+        if cancel_shift_data:
+            final_message = cancel_shift_data.final_message
+        await shift.cancel(final_message)
+        await self.__shift_repository.update(id, shift)
+        requests_to_update = []
+        users_to_update = []
+        for request in shift.requests:
+            if request.status == Request.Status.PENDING:
+                request.status = Request.Status.DECLINED.value
+                requests_to_update.append(request)
+        users = await self.__user_repository.get_users_by_shift_id(shift.id)
+        for user in users:
+            if user.status == User.Status.PENDING:
+                user.status = User.Status.DECLINED.value
+                users_to_update.append(user)
+        await self.__request_repository.update_all(requests_to_update)
+        await self.__request_repository.update_all(users_to_update)

```

Comment:
Мне казалось, что вчера оставлял такой комментарий. Исчез куда-то.
Давай разделим операции над сущностями.
```suggestion
        requests_to_update = []
        for request in shift.requests:
            ...
        await self.__request_repository.update_all(requests_to_update)

        users = await self.__user_repository.get_users_by_shift_id(shift.id)
        users_to_update = []
        for user in users:
            ...
        await self.__request_repository.update_all(users_to_update)
```